aten/src/ATen/core/boxing/impl/boxing.h

```diff
@@ -7,73 +7,252 @@
+    std::is_same<c10::ArrayRef<at::Dimname>, std::decay_t<Args>>...,
+    std::is_same<c10::optional<c10::ArrayRef<at::Dimname>>, std::decay_t<Args>>...
+  >;
 
 template<class Result, class... Args>
-Result boxAndCallBoxedFunc(KernelFunction::InternalBoxedKernelFunction* boxed_kernel_func, OperatorKernel* functor, const OperatorHandle& opHandle, Args... args, std::enable_if_t<!supports_boxing<Result, Args...>::value, int> = 0) {
-  // Some kernels don't need to actually box, and don't return.  If that's the
-  // case, just call them anyway without a stack.  These special cases can be
-  // removed once we support boxing everything.
-  // See Note [named_not_supported_kernel]
-  if (boxed_kernel_func == &named_not_supported_kernel) {
-    named_not_supported_kernel(functor, opHandle, nullptr);  // does not return
+struct BoxedKernelWrapper<Result(Args...), std::enable_if_t<has_dimname_arg<Args...>::value, void>> {
+  static Result call(KernelFunction::InternalBoxedKernelFunction*, OperatorKernel*, const OperatorHandle&, Args... args) {
+    TORCH_INTERNAL_ASSERT(false, "Call to a boxed kernel with unboxable parameter type at::Dimname.");
   }
+};
 
-  TORCH_INTERNAL_ASSERT(false, "Tried to call KernelFunction::call() for a kernel that only has a boxed kernel and doesn't support calling from an unboxed API yet.");
-}
+// at::Quantizer
+template <class... Args>
+using has_quantizer_arg =
+  guts::disjunction<
+    std::is_same<at::Quantizer, std::decay_t<Args>>...,
+    std::is_same<c10::intrusive_ptr<at::Quantizer>, std::decay_t<Args>>...
+  >;
 
-// SFINAE version for ops with returns
 template<class Result, class... Args>
-std::enable_if_t<supports_boxing<Result, Args...>::value && !std::is_same<void, Result>::value, Result>
-boxAndCallBoxedFunc(KernelFunction::InternalBoxedKernelFunction* boxed_kernel_func, OperatorKernel* functor, const OperatorHandle& opHandle, Args... args) {
-  // TODO Reuse stack vector instead of allocating?
-  torch::jit::Stack stack;
-  torch::jit::push(stack, std::forward<Args>(args)...);
+struct BoxedKernelWrapper<Result(Args...), std::enable_if_t<has_quantizer_arg<Args...>::value, void>> {
+  static Result call(KernelFunction::InternalBoxedKernelFunction*, OperatorKernel*, const OperatorHandle&, Args... args) {
+    TORCH_INTERNAL_ASSERT(false, "Unboxed call to a boxed kernel with unboxable parameter type at::Quantizer.");
+  }
+};
 
-  (*boxed_kernel_func)(functor, opHandle, &stack);
+// 2. Supported signatures, other than ref-passing.
+//
+template<class Result, class... Args>
+struct BoxedKernelWrapper<
+  Result(Args...),
+  std::enable_if_t<can_box_all<Args...>::value && can_unbox<Result>::value, void>
+> {
+  static Result call(
+    KernelFunction::InternalBoxedKernelFunction* boxed_kernel_func,
+    OperatorKernel* functor,
+    const OperatorHandle& opHandle,
+    Args... args
+  ) {
+    // TODO Reuse stack vector instead of allocating?
+    torch::jit::Stack stack;
```

Comment:
as (very) recently discussed in the group chat, we might want to change this to

```cpp
torch::jit::Stack stack {std::forward<Args>(args)...};
```

to make sure `reserve()` is called correctly and the vector doesn't grow incrementally.

Same for the other implementations below.

Reply:
doing this provoked a use-before-inference type error in the `if_constexpr` following that had no obvious cause or solution... so a) I just reserved the capacity and b) it reinforced my feelings about if_constexpr being kind of fragile and difficult

Reply:
for posterity: turns out the problem is simply `IValue` not having a constructor for `TensorOptions`.